src/migrate.c

```diff
@@ -0,0 +1,390 @@
+        if (sr->start_slot <= slot && sr->end_slot >= slot) {
+            if (sr->migration_session_key == migration_session_key) {
+                LOG_VERBOSE("migration_session keys matched");
+                return true;
+            } else {
+                LOG_VERBOSE("migration_session keys didn't match, got %llu expected %llu", migration_session_key, sr->migration_session_key);
+                return false;
+            }
+        }
+    }
+
+    LOG_VERBOSE("didn't find shardgroup slot range for this");
+    return false;
+}
+
+static int validSlotTerm(RedisRaftCtx *rr, int slot, raft_term_t term)
+{
+    ShardingInfo *si = rr->sharding_info;
+
+    if (si->max_importing_term[slot] <= term) {
+        si->max_importing_term[slot] = term;
+        return 1;
+    }
+
+    return 0;
+}
+
+static int validSlot(RedisRaftCtx *rr, int slot)
+{
+    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
+
+    if (sg && sg->local) {
+        return 1;
+    }
+
+    return 0;
+}
+
+void importKeys(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    RedisModule_Assert(entry->type == RAFT_LOGTYPE_IMPORT_KEYS);
+
+    RaftReq *req = entry->user_data;
+
+    ImportKeys import_keys = {0};
+    RedisModule_Assert(RaftRedisDeserializeImport(&import_keys, entry->data, entry->data_len) == RR_OK);
+    RedisModule_Assert(import_keys.num_keys > 0);
+
+    // FIXME: validate no cross slot migration at append time
+    int slot = keyHashSlotRedisString(import_keys.key_names[0]);
+
+    if (!validSlot(rr, slot)) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR not an importing slot");
+        }
+        goto exit;
+    }
+
+    if (!validSlotMigrationSessionKey(rr, slot, import_keys.migration_session_key)) {
+        LOG_DEBUG("ignoring import keys as incorrect migration session key");
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR invalid migration_session_key");
+        }
+        goto exit;
+    }
+
+    if (!validSlotTerm(rr, slot, import_keys.term)) {
+        // this is expected situation (suspended old leader)
+        LOG_DEBUG("ignoring import keys as old term");
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR invalid term");
+        }
+        goto exit;
+    }
+
+    /* 2  'static' strings need for restore command */
+    RedisModuleString *zero = RedisModule_CreateString(rr->ctx, "0", strlen("0")); /* ttl of zero */
+    RedisModuleString *replace = RedisModule_CreateString(rr->ctx, "REPLACE", strlen("REPLACE")); /* overwrite on import */
+
+    for (size_t i = 0; i < import_keys.num_keys; i++) {
+        RedisModuleString * temp[4];
+        temp[0] = import_keys.key_names[i];
+        temp[1] = zero;
+        temp[2] = import_keys.key_serialized[i];
+        temp[3] = replace;
+
+        enterRedisModuleCall();
+        RedisModuleCallReply *reply;
+        RedisModule_Assert((reply = RedisModule_Call(rr->ctx, "restore", "v", temp, 4)) != NULL);
```

Comment:
```suggestion
        reply = RedisModule_Call(rr->ctx, "restore", "v", temp, 4);
        RedisModule_Assert(reply != NULL);
```
`RedisModule_Assert()` is a runtime check but traditionally, `assert()` is not part of release builds. So, having a function in the assert suggests that it's not a critical function, in other words, it's just there for testing. So, I suggest moving it out.